File: clients/haskell/src/Model/Action.hs
module Model.Action where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)
import Data.Int
import Model.ActionType (ActionType)

-- | Action unit is currently performing
data Action = Action {
    -- | Tick when the action will be finished
    finishTick :: Int32,
    -- | Type of the action
    actionType :: ActionType }
    deriving Show

instance Trans Action where
    read = do
        finishTick <- Trans.read
        actionType <- Trans.read
        return Action {
            finishTick,
            actionType }
    
    write Action {
        finishTick,
        actionType } = do
            Trans.write finishTick
            Trans.write actionType

File: clients/rust/src/model/action.rs
use super::*;

/// Action unit is currently performing
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Action {
    /// Tick when the action will be finished
    pub finish_tick: i32,
    /// Type of the action
    pub action_type: model::ActionType,
}

impl trans::Trans for Action {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.finish_tick.write_to(writer)?;
        self.action_type.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let finish_tick: i32 = trans::Trans::read_from(reader)?;
        let action_type: model::ActionType = trans::Trans::read_from(reader)?;
        Ok(Self {
            finish_tick,
            action_type,
        })
    }
}